File: There is no Spoon - Episode 1/Solution.py
import sys
import math
import random

#My method for simplifying debugging
def error(s):
    print >> sys.stderr, s

# Don't let the machines win. You are humanity's last hope...

width = int(raw_input())  # the number of cells on the X axis
height = int(raw_input())  # the number of cells on the Y axis
grid = []
for i in xrange(height): #Get grid
    line = raw_input()  # width characters, each either 0 or .
    grid.insert(i, line)

nodes= []

for y in xrange(len(grid)): #Sort grid by putting in nodes a list a set of sublists ordered by their y-axis
    count = 0
    row = []
    for i in grid[y]:
        if i=='.':
            row.append(".")
            count+=1
            continue
        x = count
        row.append(str(x)+" "+str(y))
        count+=1
    nodes.append(row)

# Three coordinates: a node, its right neighbor, its bottom neighbor

nodes = [x for x in nodes if x != []] #Remove empty rows

toprint = [] #List where I store my answers

for y, line in enumerate(nodes):#Get lines and y point of each line
    if line!=[]: #Is line empty?
        for x, node in enumerate(line): #Get each node and the x point of the node
            if node != ".":
                string = str(x)+" "+str(y) #Starts the first part of the solution
                xcounter = 0 #Variable for how much I need to go over to find the next node horizontally
                ycounter = 0 #Variable for how much I need to go over to find the next node vertically
                
                while len(nodes[y])-1>x+xcounter and (nodes[y][x+xcounter]=="."or nodes[y][x+xcounter]==node):
                    #If we are not on the last x of the row and (the node where the xcounter points at is a dot or its my current node
                    #Then add 1 to xcounter
                    xcounter+=1
                xcounter = xcounter if nodes[y][x+xcounter]!="." else 0 #If the point the xcounter points at is a dot then set it to 0
                if xcounter==0: #Check if xcounter is 0, if it is...
                    string+=" -1 -1" #Add to our solution that there is no node to the right of the current node in the loop
                else: #If xcounter is not 0...
                    #Add to our solution the node where xcounter points at
                    string+=" "+str(nodes[y][x+xcounter][0])+" "+str(nodes[y][x+xcounter][2])
                
                while len(nodes)-1>y+ycounter and (nodes[y+ycounter][x]=="." or nodes[y+ycounter][x]==node):
                    #If we are not on the last row and (the node where the ycounter points at is a dot or its my current node)
                    #Then add 1 to ycounter
                    ycounter+=1
                ycounter = ycounter if nodes[y+ycounter][x]!="." else 0 #If the point the ycounter points at is a dot then set it to 0
                
                if ycounter==0: #Check if ycounter is 0, if it is...
                    string+=" -1 -1" #Add to our solution there is no node to the bottom of the current node in the loop
                else: #If ycounter is not 0
                    #Add to our solution the node where ycounter points at
                    string+=" "+str(nodes[y+ycounter][x][0])+" "+str(nodes[y+ycounter][x][2])
                toprint.append(string) #Add our solution to my list of answers
for i in toprint: #Loop through every answer
    print i #Print each answer

File: There is no Spoon - Episode 1/Solution.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

void error(string s){
    cerr << s << endl;
}
void print(string s){
    cout << s << endl;
}
/**
 * Don't let the machines win. You are humanity's last hope...
 **/
int main()
{
    int width; // the number of cells on the X axis
    cin >> width; cin.ignore();
    int height; // the number of cells on the Y axis
    cin >> height; cin.ignore();
    
    vector<string> grid;
    vector<string> row;
    int count;
    
    for (int i = 0; i < height; i++) {
        string line; // width characters, each either 0 or .
        getline(cin, line);
        grid.push_back(line);
    }
    vector<vector<string>> nodes;
    
    for (int y = 0; y<grid.size(); y++){
        count = 0;
        row.clear();
        for (char &c : grid[y]){
            if (c=='.'){
                row.push_back(".");
                count++;
                continue;
            }
            row.push_back(to_string(count)+" "+to_string(y));
            count++;
        }
        nodes.push_back(row);
    }

    // Three coordinates: a node, its right neighbor, its bottom neighbor
    for (int y=0; y<nodes.size(); y++){
        vector<string> line = nodes[y];
        for (int x=0; x<line.size(); x++){
            string node=line[x];
            if (node.compare(".")!=0) {
                string solution = to_string(x)+" "+to_string(y);
                int xcounter = 0;
                int ycounter = 0;
                
                while (nodes[y].size()-1>x+xcounter && (nodes[y][x+xcounter].compare(".")==0 
                || nodes[y][x+xcounter].compare(node)==0) ){
                    xcounter++;
                }
                xcounter = nodes[y][x+xcounter].compare(".")==0? 0 : xcounter;
                if (xcounter==0){
                    solution+=" -1 -1";
                } else {
                    solution+=" "+string(1,nodes[y][x+xcounter][0])+" "+string(1,nodes[y][x+xcounter][2]);
                }
                
                while (nodes.size()-1>y+ycounter && (nodes[y+ycounter][x].compare(".")==0 || nodes[y+ycounter][x].compare(node)==0)){
                    ycounter++;
                }
                ycounter = nodes[y+ycounter][x].compare(".")==0? 0:ycounter;
                if (ycounter==0){
                    solution+=" -1 -1";
                } else {
                    solution+=" "+string(1,nodes[y+ycounter][x][0])+" "+string(1,nodes[y+ycounter][x][2]);
                }
                print(solution);
            }
        }
    }
}
